### PR TITLE
[pytx] De-Interface FetchDelta, impl details moved to API

### PR DESCRIPTION
**Pull request by Dcallies**
Summary
---------

As a final effort to simplify the SignalExchange interfaces, this moves the details on merging updates into the SignalExchangeAPI itself.

This also has some dangling requests from #1017 that I moved here to avoid one pass there.

Test Plan
---------

py.test && mypy

### CODE REVIEW

File: python-threatexchange/threatexchange/fetcher/fetch_api.py
Comment (Dcallies) [inline python-threatexchange/threatexchange/fetcher/fetch_api.py:134]: This is basically the key part of the PR

File: python-threatexchange/threatexchange/fetcher/fetch_state.py
Comment (Dcallies) [inline python-threatexchange/threatexchange/fetcher/fetch_state.py:83]: This leaked from another set of commits, but worth keeping in I think

**Review by BarrettOlson — APPROVED** [review]
Approving to unblock. updates seems like steps in the right direction.

File: python-threatexchange/threatexchange/fetcher/apis/tests/test_stopncii.py
Comment (schatten) [inline python-threatexchange/threatexchange/fetcher/apis/tests/test_stopncii.py:24]: non-blocking. We've used [freezegun](https://github.com/spulec/freezegun) for moving through time in HMA. If the testing gets any more complex, freezegun's API is much more readable and does not rely on magic numbers.

File: python-threatexchange/threatexchange/fetcher/apis/tests/test_stopncii.py
Comment (schatten) [inline python-threatexchange/threatexchange/fetcher/apis/tests/test_stopncii.py:30]: Do you make actual API calls here? Can this test start failing because of the underlying database receiving updates?

**Review by schatten — APPROVED** [review]
Looks good. Comments inline. The only open question I'd like answered before you merge is this:

- If for large datasets, we MUST move the merge impl into the FetchStateStore implementation, why should the API interface support merging at all?
  - Why not make that a responsibility of the store itself?

Also, a strongly held nit about not editing function values if you are also returning said value.

**Comment by Dcallies** [comment]
> If for large datasets, we MUST move the merge impl into the FetchStateStore implementation, why should the API interface support merging at all? 

Because I posit that almost no database has this problem currently. Hashes are generally small. Providing in-memory as a fallback means that a fallback always exists, which is enough to prototype and test, which is the main usecase I'm targeting with the `naive_` functions. If you can prove the value, you can justify the effort of building a store.

> Why not make that a responsibility of the store itself?
My goal was to reduce the total number of interfaces that need to prototype to 1 - SignalAPI. If I did this, it would increase to 2, and also whatever storage you choose might not work in all cases. 

---

I am generally encouraging all interfaces to be built off dataclass, and did at least prototype structured storage that would be needed. There's also a potential solution that I debated, which is enforcing that all updates must be keyed - if that's the case, it always allow partitioning by the key, and so always has an approach to incremental storage. I chose not to take that step. 

However, that step might be worth taking because it does have a translation for key just being a constant, which then puts the burden on the API for not making individual values larger than what is likely to be available to memory.

File: python-threatexchange/threatexchange/fetcher/apis/tests/test_stopncii.py
Comment (Dcallies) [reply python-threatexchange/threatexchange/fetcher/apis/tests/test_stopncii.py:24]: I might attempt to convince you in the other direction - I've found the py.test fixture functionality pretty usable, and just has a short barrier to understanding it.

File: python-threatexchange/threatexchange/fetcher/fetch_api.py
Comment (Dcallies) [reply python-threatexchange/threatexchange/fetcher/fetch_api.py:312]: I considered adding them as mixins, but I've confused myself in the past by ordering methods incorrectly when using mixins. Adding the diamond explicitly provides a consistent resolution order.

File: python-threatexchange/threatexchange/fetcher/fetch_api.py
Comment (Dcallies) [reply python-threatexchange/threatexchange/fetcher/fetch_api.py:301]: It's explained you can do so in the interface comment.

**Pull request merged**
